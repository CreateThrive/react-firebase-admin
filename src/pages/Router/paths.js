export default {
  ROOT: '/',
  LOGIN: '/login',
  USERS: '/users',
  ADD_USER: '/users/new',
  MODIFY_USER: '/users/:id',
  PROFILE: '/profile',
  RESET_PASSWORD: '/recover-password',
  SECTION: '/section',
  SUBMENU_1: '/submenu1',
  SUBMENU_2: '/submenu2',
  TEAMS: '/teams',
};
