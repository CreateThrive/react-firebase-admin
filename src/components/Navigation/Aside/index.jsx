import React, { useState } from 'react';
import { useSelector, shallowEqual } from 'react-redux';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { useFormatMessage } from 'hooks';
import paths from 'pages/Router/paths';
import NavLink from '../Link';
import classes from './Aside.module.scss';

export const SubMenu = ({ label, children }) => {
  const [active, setActive] = useState(false);

  return (
    <li className={classNames({ 'is-active': active })}>
      <a
        exact-active-class="is-active"
        className="has-icon has-dropdown-icon"
        onClick={() => setActive(!active)}
      >
        <span className="icon">
          <i className="mdi mdi-view-list" />
        </span>
        <span className="menu-item-label">{label}</span>
        <div className="dropdown-icon">
          <span className="icon">
            <i
              className={classNames(
                'mdi',
                { 'mdi-minus': active },
                { 'mdi-plus': !active }
              )}
            />
          </span>
        </div>
      </a>
      <ul>{children}</ul>
    </li>
  );
};

SubMenu.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.string.isRequired
};

const Aside = ({ handleMobileToggle }) => {
  const { isAdmin } = useSelector(
    state => ({
      isAdmin: state.auth.userData.isAdmin
    }),
    shallowEqual
  );

  const usersMessage = useFormatMessage('Aside.users');

  return (
    <aside className="aside is-placed-left is-expanded">
      <Link to={paths.ROOT} className="aside-tools">
        <div className="aside-tools-label">
          <span>
            <b>React</b> Firebase
          </span>
        </div>
      </Link>
      <div className="menu is-menu-main">
        <ul className="menu-list">
          <li>
            <NavLink
              to={paths.ROOT}
              className="has-icon"
              onClick={handleMobileToggle}
            >
              <span className="icon">
                <i className="mdi mdi-home" />
              </span>
              <span className="menu-item-label">
                {useFormatMessage('Aside.home')}
              </span>
            </NavLink>
          </li>
          {isAdmin && (
            <li>
              <NavLink
                to={paths.USERS}
                className="has-icon"
                onClick={handleMobileToggle}
              >
                <span className="icon">
                  <i className="mdi mdi-account-supervisor" />
                </span>
                <span className="menu-item-label">{usersMessage}</span>
              </NavLink>
            </li>
          )}
          <SubMenu label={useFormatMessage('Aside.dropdownMenu')}>
            <li>
              <NavLink
                className={classes.submenuLink}
                to={paths.SUBMENU_1}
                onClick={handleMobileToggle}
              >
                <span>{useFormatMessage('Aside.submenu1')}</span>
              </NavLink>
            </li>
            <li>
              <NavLink
                className={classes.submenuLink}
                to={paths.SUBMENU_2}
                onClick={handleMobileToggle}
              >
                <span>{useFormatMessage('Aside.submenu2')}</span>
              </NavLink>
            </li>
          </SubMenu>
        </ul>
      </div>
    </aside>
  );
};

Aside.propTypes = {
  handleMobileToggle: PropTypes.func.isRequired
};

export default Aside;
