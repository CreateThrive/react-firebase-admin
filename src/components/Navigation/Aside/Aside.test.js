import React from 'react';

import paths from 'pages/Router/paths';
import Aside, { SubMenu } from '.';
import NavLink from '../Link';

describe('<Aside /> rendering', () => {
  const onHandler = jest.fn();

  it('should render without crashing', () => {
    const component = shallowWithProviders(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(component).toMatchSnapshot();
  });

  it('should set the handleMobileToggle prop correctly', () => {
    const { component } = mountWithProviders(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(
      component
        .find(NavLink)
        .at(1)
        .prop('onClick')
    ).toEqual(onHandler);

    expect(
      component
        .find(NavLink)
        .at(0)
        .prop('onClick')
    ).toEqual(onHandler);

    expect(
      component
        .find(NavLink)
        .at(2)
        .prop('onClick')
    ).toEqual(onHandler);
  });

  it('should not render the /users link if it the user is not an admin', () => {
    const { component } = mountWithProviders(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: false
        }
      }
    });

    expect(
      component.contains(
        <NavLink to={paths.USERS} onClick={onHandler}>
          Users
        </NavLink>
      )
    ).toBeFalsy();
  });

  it('should render the /users link if it the user is an admin', () => {
    const { component } = mountWithProviders(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(
      component.contains(
        <NavLink to={paths.USERS} onClick={onHandler}>
          Users
        </NavLink>
      )
    ).toBeFalsy();
  });

  it('should render the <SubMenu /> component if the user is an admin', () => {
    const { component } = mountWithProviders(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(component.exists(SubMenu)).toBeTruthy();
  });

  it('should render the <SubMenu /> component if the user is not an admin', () => {
    const { component } = mountWithProviders(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: false
        }
      }
    });

    expect(component.exists(SubMenu)).toBeTruthy();
  });
});
