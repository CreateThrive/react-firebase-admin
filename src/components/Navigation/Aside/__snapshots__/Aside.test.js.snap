// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Aside /> rendering should render without crashing 1`] = `
<DocumentFragment>
  <aside
    class="aside is-placed-left is-expanded"
  >
    <a
      class="aside-tools"
      href="/"
    >
      <div
        class="aside-tools-label"
      >
        <span>
          <b>
            React
          </b>
           Firebase
        </span>
      </div>
    </a>
    <div
      class="menu is-menu-main"
    >
      <ul
        class="menu-list"
      >
        <li>
          <a
            aria-current="page"
            class="has-icon is-active"
            href="/"
          >
            <span
              class="icon"
            >
              <i
                class="mdi mdi-home"
              />
            </span>
            <span
              class="menu-item-label"
            >
              Home
            </span>
          </a>
        </li>
        <li>
          <a
            class="has-icon"
            href="/users"
          >
            <span
              class="icon"
            >
              <i
                class="mdi mdi-account-supervisor"
              />
            </span>
            <span
              class="menu-item-label"
            >
              Users
            </span>
          </a>
        </li>
        <li>
          <a
            class="has-icon"
            href="/teams"
          >
            <span
              class="icon"
            >
              <i
                class="mdi mdi-account-group"
              />
            </span>
            <span
              class="menu-item-label"
            >
              Teams
            </span>
          </a>
        </li>
        <li
          class=""
        >
          <a
            class="has-icon has-dropdown-icon"
            exact-active-class="is-active"
          >
            <span
              class="icon"
            >
              <i
                class="mdi mdi-view-list"
              />
            </span>
            <span
              class="menu-item-label"
            >
              Dropdown Menu
            </span>
            <div
              class="dropdown-icon"
            >
              <span
                class="icon"
              >
                <i
                  class="mdi mdi-plus"
                />
              </span>
            </div>
          </a>
          <ul>
            <li>
              <a
                class="submenuLink"
                href="/submenu1"
              >
                <span>
                  Submenu 1
                </span>
              </a>
            </li>
            <li>
              <a
                class="submenuLink"
                href="/submenu2"
              >
                <span>
                  Submenu 2
                </span>
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </aside>
</DocumentFragment>
`;
